racket/collects/racket/private/kw.rkt

```diff
@@ -39,6 +41,139 @@
+  ;;                              (add1 b)
+  ;;                              _a2)]
+  ;;                       [z (if (#%expression #f) '3 _z)]
+  ;;                       [r _r])
+  ;;                  <body>))])
+  ;;    (let ([unpack (lambda (kws args _x _y _z _r)
+  ;;                    (let* ([has-a? (and (pair? kws)
+  ;;                                        (eq? '#:a (car kws)))]
+  ;;                           [_a (if has-a? (car args) unsafe-undefined)]
+  ;;                           [kws (if has-a? (cdr kws) kws)]
+  ;;                           [args (if has-a? (cdr args) args)]
+  ;;                           [has-b? (pair? args)]
+  ;;                           [_b (if has-b? (car args) 'b)])
+  ;;                      (core _a _b _x _y _z _r)))])
+  ;;      (make-optional-keyword-procedure
+  ;;       ...
+  ;;       ;; Entry point when at least one keyword argument is provided:
+  ;;       (case-lambda
+  ;;         [(kw args x) (unpack kw args x unsafe-undefined '3 null)]
+  ;;         [(kws args x y z . r) (unpack kws args x y z r)]
+  ;;         [(kws args x y) (unpack kws args x y '3 null)])
+  ;;       ...
+  ;;       ;; Entry point when no keywords are provided:
+  ;;       (case-lambda
+  ;;         [(x) (unpack null null x unsafe-undefined '3 null)]
+  ;;         [(x y z . r) (unpack null null x y z r)]
+  ;;         [(x y) (unpack null null x y '3 null)]))))
+  ;;
+  ;;
+  ;; Another example, illustrating a mandatory keyword argument:
+  ;;
+  ;;  (lambda (#:x x #:y [y (add1 x)]) <body>)
+  ;;
+  ;; becomes
+  ;;
+  ;;  (let ([core (lambda (_x _y)
+  ;;                (let* ([x _x]
+  ;;                       [y (if (eq? _y unsafe-undefined)
+  ;;                              (add1 x)
+  ;;                              _y)])
+  ;;                  <body>))])
+  ;;    (let ([unpack
+  ;;           (lambda (kws args)
+  ;;             (let* ([_x (car args)] ; no check needed
+  ;;                    [kws (cdr kws)]
+  ;;                    [args (cdr args)]
+  ;;                    [has-y? (pair? kws)]
+  ;;                    [_y (if has-y? (car args) unsafe-undefined)])
+  ;;               (core _x _y)))])
+  ;;      (naming-constructor
+  ;;       ...
+  ;;       (case-lambda
+  ;;         [(kws args) (unpack kw args)])
+  ;;       ...)))
+  ;;
+  ;; Finally, `(define (f ...) <body>)` or `(define f (lambda (...)
+  ;; <body>))` with keyword arguments expands to bind `f` as a macro,
+  ;; and some `_f` is bound to the expansion illustrated above, except
+  ;; that the `core` and `unpack` bindings are flattened into the
+  ;; definition context. That way, uses of the `f` macro can typically
+  ;; expand to a direct call to the corresponding `core` function,
+  ;; statically parsing the supplied keyword arguments and passing
+  ;; `unsafe-undefined` or an immediate default in place of unsupplied
+  ;; arguments. This macro-binding approach is used only when `f` has
+  ;; keyword arguments.
+
```

Comment:
It would be helpful to have a similar example for the expansion of various application forms (with/without keyword arguments; statically-known functions that have simple/complicated defaults).

Reply:
ok